Add import resolution plugin hook

SamuelRobertBond opened the pull request:
I was having issues with files with a . in the file name (e.g physics-api.component.ts) and needed some method of providing custom filepath resolution login when importing files from from a library into another project bundling the library.

SamuelRobertBond commented:
I found some issues with paths in this, currently working through it. Let me know if you have any thoughts on what I can add or change.

SamuelRobertBond commented:
Fixed the issue and moved the plugin hook call to `resolveLuaDependencyPathFromNodeModules`. I uploaded an example usage [here](https://github.com/SamuelRobertBond/tstl-import-resolution-plugin-example)

SamuelRobertBond commented:
Another idea I had for changes to the plugins would be to allow plugins definitions to be passed through the compiler options when using tstl through the api. It would be a seperate issue, but I wanted to get your thoughts on this before I start messing around with that.

This is an example of what im talking about, although I may just not be understanding the `TransformerImport`
![image](https://github.com/TypeScriptToLua/TypeScriptToLua/assets/17439332/a34e67bc-8ddf-4d37-9695-ead54db1235f)

Perryvw commented on src/transpilation/plugins.ts:
```diff
@@ -45,6 +45,8 @@ export interface Plugin {
         result: EmitFile[]
     ) => ts.Diagnostic[] | void;
+
+    onImportResolutionFailure?: (packageRoot : string, dependency : string) => string | undefined
```
A module resolution plugin is something we have long been considering, and I think a better interface would be something like:

```ts
interface ModuleResolutionRequest {
    moduleIdentifier: string; // "foo.bar.baz" from require("foo.bar.baz")
    requiringFile: string; // path of the (lua) file containing the require
    // maybe some other stuff I'm forgetting now?
    // maybe a 'defaultResolve: (ModuleResolutionRequest) => string' to call the default behavior?
    // Not sure why anyone would want that
}

moduleResolution?: (moduleToBeResolved: ModuleResolutionRequest) => string | undefined
```

Then crucially this plugin would be executed _before_ our regular module resolution process, allowing users to intercept before it even starts. If it returns undefined it will go into the default behavior. I'm just wondering if we want to somehow maybe also be able to indicate that a specific import should _not_ be emitted, but maybe that's for a future version.

SamuelRobertBond replied:
I think this interface makes more sense, it gives more control to someone writing a plugin to identify the package they wish to use via the module identifier. Ill start working on it.

Perryvw commented on src/transpilation/plugins.ts:
```diff
@@ -45,6 +45,8 @@ export interface Plugin {
         result: EmitFile[]
     ) => ts.Diagnostic[] | void;
+
+    onImportResolutionFailure?: (packageRoot : string, dependency : string) => string | undefined
```
I don't like you passing packageRoot to the outside world, that's really something internal to the tstl implementation. What is the reason you need this in your case? I'm trying to understand the usage scenario

SamuelRobertBond replied:
In my particular case, I was getting the path the package and the attempted path to the dependency and transforming that to what I needed it to be, so when the plugin. Before I made the changes to packageRoot, the path being passed to the plugin was inconsistent. This was just much more consistent for me to work with. I think the changes to the interface and location for the plugins call will resolve this issue.

Perryvw commented on src/transpilation/resolve.ts:
```diff
@@ -184,17 +186,35 @@ class ResolutionContext {
+            }
+        }
+
+        if (this.plugins != null) {
```
I don't think this should be exclusive to nodeModules paths, I think a nicer place for this would be in `resolveImport`, before the call to `resolveDependencyPath` (going with the idea I mentioned on the plugin interface above, where users can intercept these calls before we do all of this quite complicated work on the file system while we already know we won't find anything)

Perryvw commented on src/transpilation/resolve.ts:
```diff
@@ -235,6 +255,11 @@ class ResolutionContext {
             path.join(resolvedPath, "init.lua"), // lua looks for <require>/init.lua if it cannot find <require>.lua
         ];
+
+        if (resolvedPath.endsWith(".lua")) {
```
I don't think this if statement should be here at all. If the resolved path ends with .lua we do not have to search for it, because we apparently we already have a lua file path. I don't think this is possible in regular tstl module resolution, but I guess your plugin triggers this? I don't think this code path should really be possible ever.

SamuelRobertBond replied:
I put this here as a convince thing for me, since the initial problem I intend to solve was being able to specify. When I was constructing the path it I included .lua in the resolved path. This was not being found since the possible files above appends .lua to the resolve path, the path was being set to /my/path/file.lua.lua. I didn't want to just add resolvedPath itself to the possible files, because someone could intentionally provide a path to a non lua file. I think the changes you mentioned above would make this a non issue though.

SamuelRobertBond commented:
I made some of the changes requested. I didnt include `defaultResolve` yet. I also still need to write the test. Figured I'd get them out there in case you wanted to take a look. Ill probably write the test tomorrow.

Perryvw reviewed (commented):
Starting to look good, need a test still, and of course you should make sure this works for your use-case too still. Also to avoid our formatter complaining, you can run `npm run lint` and `npm run fix:prettier` locally to make sure there are no formatting or linting complaints.

SamuelRobertBond commented:
Fixed an issue I found with the path resolution and added the test specified. I ran and fixed the formatting with the linter and ran the entire test suite with everything passing. I wasn't sure if I should add a test in the plugins spec as well. I included the moduleResolution plugin for the test in the plugins folder, but I wasn't sure if that was the appropriate place for it. Let me know if there's anything else I should include in the test or anything that needs to be changed.

Perryvw commented on src/transpilation/resolve.ts:
```diff
@@ -93,6 +98,60 @@ class ResolutionContext {
+            if (plugin.moduleResolution != null) {
+                const pluginResolvedPath = plugin.moduleResolution(
+                    dependency,
+                    dependencyPath,
```
this parameter is the requiringFile, shouldn't this be `required.fileName`?

SamuelRobertBond replied:
I missed that, which make sense why I ran into the issues I did, I'll fix that ty

Perryvw commented on src/transpilation/resolve.ts:
```diff
@@ -93,6 +98,60 @@ class ResolutionContext {
+                        }
+                    }
+
+                    const isRelative = ["/", "./", "../"].some(p => pluginResolvedPath.startsWith(p));
+
+                    // // If the import is relative, always resolve it relative to the requiring file
+                    // // If the import is not relative, resolve it relative to options.baseUrl if it is set
+                    const fileDirectory = path.dirname(required.fileName);
+                    const relativeTo = isRelative ? fileDirectory : this.options.baseUrl ?? fileDirectory;
+
+                    // // Check if file is a file in the project
+                    const resolvedPath = path.join(relativeTo, pluginResolvedPath);
+                    const fileFromPath = this.getFileFromPath(resolvedPath);
```
This looks just straight up duplicate from logic in `resolveDependencyPath`, please move to a shared function they both can use

Perryvw commented on test/transpile/plugins/moduleResolution.ts:
```diff
@@ -0,0 +1,12 @@
+const plugin: tstl.Plugin = {
+    moduleResolution(moduleIdentifier, requiringFile) {
+        if (moduleIdentifier.includes("foo")) {
+            return requiringFile.replace("foo", "bar");
```
the requiring file is called main.ts (or .lua?), not foo, so I'm confused by this plugin. See also my other comment regarding the requirintFile parameter

Perryvw reviewed (approved):
Looks good to me. Thanks for your contribution!
